foxglove/extensions/pid-panel/src/PIDPanel.tsx

```diff
@@ -0,0 +1,341 @@
+        });
+      });
+    });
+
+    callService(PID_SERVICE, request);
+    handleReset();
+  };
+
+  const theme = useTheme();
+  return (
+    <ThemeProvider theme={theme}>
+      <Box m={1}>
+        {/* Error messages */}
+        {(context.callService == undefined || state.error != undefined) && (
+          <Box mb={1}>
+            {context.callService == undefined && (
+              <Alert variant="filled" severity="error">
+                Calling services is not supported by this connection.
+              </Alert>
+            )}
+            {state.error != undefined && (
+              <Alert variant="filled" severity="error">
+                {state.error.message}
+              </Alert>
+            )}
+          </Box>
+        )}
+
+        {/* PID Loop Tabs */}
+        <Tabs value={state.loop} onChange={handleLoopChange} variant="fullWidth">
+          <Tab label="Position" value={CustomMsgsPidGainConst.LOOP_POSITION} />
+          <Tab label="Velocity" value={CustomMsgsPidGainConst.LOOP_VELOCITY} />
+        </Tabs>
+
+        {/* PID Gains Table */}
+        <TableContainer>
+          <Table size="small">
+            <TableHead>
+              <TableRow>
+                <TableCell />
+                <TableCell style={{ textAlign: "center" }}>KP</TableCell>
+                <TableCell style={{ textAlign: "center" }}>KI</TableCell>
+                <TableCell style={{ textAlign: "center" }}>KD</TableCell>
+                <TableCell style={{ textAlign: "center" }}>FF</TableCell>
+              </TableRow>
+            </TableHead>
+            <TableBody>
+              {/* Loop through all axes/gain types */}
+              {Object.entries(state.pid[state.loop] ?? {}).map(([axis, gains]) => (
+                <TableRow key={axis}>
+                  {/* Axis Label */}
+                  <TableCell>{getAxisEnumName(Number(axis))}</TableCell>
+
+                  {/* Axis Gains */}
+                  {Object.values(gains).map((gain, gainType) => {
+                    const inEditMode = gainType in (state.editedGains[Number(axis)] ?? {});
+                    return (
+                      <TableCell key={gainType} sx={{ padding: "0" }}>
+                        <TextField
+                          type="text"
+                          sx={{
+                            "& input": {
+                              boxSizing: "border-box",
+                            },
+                            "& .MuiOutlinedInput-root": {
+                              "& fieldset": {
+                                // Border color is red if in edit mode, otherwise divider color
+                                borderColor: `${inEditMode ? theme.palette.error.main : theme.palette.divider}`,
+                              },
+                            },
+                          }}
+                          value={inEditMode ? undefined : gain} // Display the true gain value if not in edit mode
```

Comment:
To complete [the loop of react](https://legacy.reactjs.org/docs/forms.html#controlled-components), when `inEditMode` is `true`, value should be set to the edited value and not just `undefined`.